### PR TITLE
Adding Nginx sidecar to Rails app Helm charts.

### PR DESCRIPTION
**Pull request by nsabri1**
Adding Nginx reverse proxy to Rails apps as sidecar.

We're not 100% sure what behaviour these will need in the future. Possibly they'll be involved in serving static assets  such ad CSS, JS etc.

It will certainly be needed in some applications such asset manager, which uses ‘x-accel-redirect’.

We are not mandating the use of Nginx sidecar for every Rails app, though we believe it will be the case.

If it turns out that the Nginx sidecar have no behaviour, we may consider removing them at a later date.

Some of the reasons to why we are adding Nginx sidecar:
 
We want apps to be able to work as independently as possible.

Helm chart should be a self-contained encapsulated artefact.

Nginx sidecar will just be doing reverse proxy, static serving and possibly some x-accel-redirect stuff to s3 for the few apps that need this.

Security benefit of having Nginx sit between inbound http traffic and Unicorn.

Unicorn can't deal with slow clients efficiently.

Nginx sidecar Config:

The Nginx config itself will need to be updated to support functionality we require in the future, 

Some reference for NGINX requirements can be found here:
https://docs.google.com/document/d/1hOgkeimIn20O5zrD0wTflap1_DTVatb7NOnb223NhSE/
https://trello.com/c/Lkb8N313/381-epic-nginx-and-varnish-changes-needed

Values naming changes:

Rails application image to: appImage and 
Application container port to: appPort

Nginx sidecar image to - nginxImage
Nginx container portto: nginxPort

### CODE REVIEW

**Comment by sengi** [comment]
Having a proper look now but in the meantime please fix the lint errors ;)

File: helm/content-store/templates/configmap.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-nginx-conf
+data:
+  nginx.conf: |-
+    user  nginx;
+    worker_processes  1;
```
Comment (sengi) [inline helm/content-store/templates/configmap.yaml:8]: This is already the default and we're not likely to need to change it, so I'd probably omit it.
```suggestion
```

File: helm/content-store/values.yaml [hunk 2]
```diff
@@ -10,17 +10,21 @@ service:
...
+appImage:
   repository: 172025368201.dkr.ecr.eu-west-1.amazonaws.com/content-store
   pullPolicy: IfNotPresent
   tag: latest
 
-containerPort: 80
+nginxImage:
+  repository: nginx
+  pullPolicy: IfNotPresent
+  tag: latest
```
Comment (sengi) [inline helm/content-store/values.yaml:24]: I'd recommend using stable as a starting point. Latest can be the development branch with nginx: see their [supported tags](https://hub.docker.com/_/nginx). (Ultimately we might want to pin the versions and have something like Dependabot update them in a managed way, but that's way in the future.)
```suggestion
  tag: stable
```

File: helm/content-store/values.yaml [hunk 3]
```diff
@@ -10,17 +10,21 @@ service:
...
-image:
+appImage:
   repository: 172025368201.dkr.ecr.eu-west-1.amazonaws.com/content-store
   pullPolicy: IfNotPresent
   tag: latest
 
-containerPort: 80
+nginxImage:
+  repository: nginx
+  pullPolicy: IfNotPresent
```
Comment (kerin) [inline helm/content-store/values.yaml:23]: Something to be aware of here - if `imageTag` is `latest` and `pullPolicy` is `IfNotPresent` then Kubelet on the node will pull the image the first time then never pull it again, unless either the tag or pull policy is changed at a later date. This will also apply to other unrelated Pods or Deployments that land on the same worker that also use the same `image:tag` and `pullPolicy: IfNotPresent` combo, as from Kubernetes' point an image tagged with `latest` is indeed already `Present`.

This is one reason why it's recommended to [not use `latest`](https://kubernetes.io/docs/concepts/containers/images/#image-pull-policy), because it's non-deterministic; or, to specify `imagePullPolicy: Always`, so that the latest image is always pulled on deploy. There are pros and cons to both pinning a specific tag and always pulling `latest`, but generally I'd lean towards being specific, so that we don't get an unexpected breaking change from a new major version one day.

tl;dr - I'd change `tag` to `1.21.1`

Comment (sengi) [inline helm/content-store/values.yaml:23]: If you switch from `latest` to `stable` (or anything else for that matter) then you'll no longer need to specify this (i.e. it'll default to `IfNotPresent`).
```suggestion
```

**Review by kerin — CHANGES_REQUESTED** [review]
There's one sneaky issue lurking here ready to surprise us in the future, otherwise LGTM 👍

File: helm/content-store/templates/configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,38 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-nginx-conf
+data:
+  nginx.conf: |-
+    user  nginx;
```
Comment (sengi) [inline helm/content-store/templates/configmap.yaml:7]: Formatting nits (sorry 🙈) - the indentation of the values looks a bit funky here (through to the rest of the file). It looks like they were once lined up with spaces but have since gone wonky 😁

I'd probably just go with one space between the key and the value throughout and save the faff of trying to line things up (I don't think making all the values line up helps readability and it leads to unnecessary diffs and merge conflicts if it's enforced).

File: helm/content-store/values.yaml (hunk 3 above)
Comment (sengi) [reply helm/content-store/values.yaml:23]: Snap :) If you go with a specific version then we'll need an update mechanism though (which we'll want eventually but it'll probably be a while). We could go with `stable` in the interim so at least we'll get security updates.

Comment (sengi) [reply helm/content-store/values.yaml:23]: Sorry I'm being an idiot - that'll have the same issue. The interim thing until we have an update mechanism would be to keep pullPolicy: Always in conjunction with using `stable`.

Comment (sengi) [reply helm/content-store/values.yaml:23]: Sorry didn't think this through properly - if we're using any label that can move around we don't really want IfNotPresent because there's no way for it to know to re-pull when the label moves.

Comment (kerin) [reply helm/content-store/values.yaml:23]: Yeah, either approach is fine in this case I think, particularly as a specific tag can always be provided at `helm install` time - a default of `stable` with `Always` seems reasonable 👍

Comment (sengi) [reply helm/content-store/values.yaml:23]: So it's either `pullPolicy: Always` and `stable` or `1.20.1` and no/default PullPolicy and a big scary TODO to figure out a Dependabot-like thing to update the pinned versions.

Comment (kerin) [reply helm/content-store/values.yaml:23]: This is very much something for another day, but as any Docker tag can arbitrarily move around the only 100% safe usage of `IfNotPresent` is when using a container's image digest (`image@sha256:`) rather than a tag - definitely not something we need to get into now though.

Comment (sengi) [reply helm/content-store/values.yaml:23]: > particularly as a specific tag can always be provided at helm install time
Ah yeah good point

Comment (kerin) [reply helm/content-store/values.yaml:23]: Yeah go with `stable / Always` - it's a reasonable default when nothing specific has been requested at `helm install` time.
